lib/Dialect/CGGI/IR/CGGIOps.cpp

```diff
@@ -8,17 +8,38 @@
 #include "mlir/include/mlir/IR/ValueRange.h"          // from @llvm-project
 #include "mlir/include/mlir/Support/LLVM.h"           // from @llvm-project
 #include "mlir/include/mlir/Support/LogicalResult.h"  // from @llvm-project
+#include "mlir/include/mlir/Transforms/GreedyPatternRewriteDriver.h"  // from @llvm-project
 
 namespace mlir {
 namespace heir {
 namespace cggi {
 
 ValueRange Lut2Op::getLookupTableInputs() { return ValueRange{getB(), getA()}; }
 
+LogicalResult Lut2Op::canonicalize(Lut2Op op, PatternRewriter &rewriter) {
+  static const int32_t array2[] = {1, 1};
```

Comment:
Nice! This is a great start - actually the LUT 2 assembles two bits into a single ciphertext input with shifting and adding (https://github.com/google/heir/blob/9587f902f1f755315826db847e5be21cefbba5c0/lib/Dialect/CGGI/IR/CGGIOps.td#L97-L98), so this would actually be `{1, 2}`, and LUT 3 would be `{1, 2, 4}`

Reply:
aha - I see